py/server/deephaven/remote.py

```diff
@@ -0,0 +1,135 @@
+from typing import Dict
+
+import jpy
+
+from deephaven import DHError, uri
+from deephaven._wrapper import JObjectWrapper
+from deephaven.table import Table
+
+_JURI = jpy.get_type("java.net.URI")
+_JClientConfig = jpy.get_type("io.deephaven.client.impl.ClientConfig")
+_JSSLConfig = jpy.get_type("io.deephaven.ssl.config.SSLConfig")
+_JTrustCustom = jpy.get_type("io.deephaven.ssl.config.TrustCustom")
+_JDeephavenTarget = jpy.get_type("io.deephaven.uri.DeephavenTarget")
+_JBarrageSession = jpy.get_type("io.deephaven.client.impl.BarrageSession")
+_JSharedId = jpy.get_type("io.deephaven.client.impl.SharedId")
+_JBarrageTableResolver = jpy.get_type("io.deephaven.server.uri.BarrageTableResolver")
+
+_session_cache: Dict[str, BarrageSession] = {}  # use WeakValueDictionary to avoid memory leak?
+_remote_session_lock = threading.Lock()
+
+
+def barrage_session(host: str,
+                    port: int = 10000,
+                    auth_type: str = "Anonymous",
+                    auth_token: str = "",
+                    use_tls: bool = False,
+                    tls_root_certs: bytes = None,
+                    ) -> BarrageSession:
+    """Returns a Deephaven gRPC session to a remote server if a cached session is available; otherwise, creates a new
+    session.
+
+    Note: client authentication is not supported yet.
+
+    Args:
+        host (str): the host name or IP address of the Deephaven server.
+        port (int): the port number that the remote Deephaven server is listening on, default is 10000.
+        auth_type (str): the authentication type string, can be "Anonymous', 'Basic", or any custom-built
+            authenticator in the server, such as "io.deephaven.authentication.psk.PskAuthenticationHandler",
+            default is 'Anonymous'.
+        auth_token (str): the authentication token string. When auth_type is 'Basic', it must be
+            "user:password"; when auth_type is "Anonymous', it will be ignored; when auth_type is a custom-built
+            authenticator, it must conform to the specific requirement of the authenticator
+        use_tls (bool): if True, use a TLS connection.  Defaults to False
+        tls_root_certs (bytes): PEM encoded root certificates to use for TLS connection, or None to use system defaults.
+             If not None implies use a TLS connection and the use_tls argument should have been passed
+             as True. Defaults to None
+
+    Returns:
+        a Deephaven Barrage session
+
+    Raises:
+        DHError
+    """
+    try:
+        j_barrage_session_factory_client = uri.resolve(
+            "dh:///app/io.deephaven.server.barrage.BarrageSessionFactoryClient/field/instance")
+        if tls_root_certs and not use_tls:
+            raise DHError(message="tls_root_certs is provided but use_tls is False")
+
+        target_uri = f"{host}:{port}"
+        if use_tls:
+            target_uri = f"dh://{target_uri}"
+        else:
+            target_uri = f"dh+plain://{target_uri}"
```

Comment:
I wonder if we should just use the uri format here? I know the params might more closely match what you are doing on pydeephaven side, but there is a nice simplicity in a single DH URI string, like `dh://remote_host` (as opposed to `host="remote_host", port=443, use_tls=True`):

```java
    /**
     * Parses the {@code targetUri} into a Deephaven target.
     *
     * <p>
     * The valid formats include {@code dh://host}, {@code dh://host:port}, {@code dh+plain://host}, and
     * {@code dh+plain://host:port}.
     *
     * @param targetUri the target URI
     * @return the Deephaven target
     */
    public static DeephavenTarget of(URI targetUri) {
```